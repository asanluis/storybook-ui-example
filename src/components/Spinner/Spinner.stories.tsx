import React from 'react';
import { ComponentMeta, Story } from '@storybook/react';

import Spinner from './index';
import {SpinnerProps} from "./Spinner.types";

export default {
  title: 'Components/Spinner',
  component: Spinner,
  parameters: {
    controls: {
    },
    default: 'Default',
  },
  argTypes: {
    rotate: { control: 'number' },
    delay: { control: 'number' }
  },
} as ComponentMeta<typeof Spinner>;


const Template: Story<SpinnerProps> = ({
  ...args
}) => (
    <Spinner {...args} />
);

export const TooFast = Template.bind({});
TooFast.args = {
  rotate: 30,
  delay: 10,
};

export const Default = Template.bind({});
Default.args = {
  rotate: 30,
  delay: 0.1,
};
