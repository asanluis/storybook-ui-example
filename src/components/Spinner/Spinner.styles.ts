import styled, { keyframes } from 'styled-components';

export const CircleContainer = styled.div`
  margin: 100px auto;
  width: 40px;
  height: 40px;
  position: relative;
`;

const CircleDelay = keyframes`
  0%,
  39%,
  100% {
    opacity: 0;
  }
  40% {
    opacity: 1;
  }
`;

interface SpinnerProps {
    rotate?: string;
    delay?: string;
}
export const OneCircle = styled.div<SpinnerProps>`
  width: 100%;
  height: 100%;
  position: absolute;
  transform: rotate(${props => props.rotate});
  left: 0;
  top: 0;
  &::before {
    background-color: ${props => props.theme.colors.main};;
    content: '';
    display: block;
    margin: 0 auto;
    width: 15%;
    height: 15%;
    border-radius: 100%;
    animation: ${CircleDelay} 1.2s infinite ease-in-out both;
    animation-delay: ${props => props.delay};
  }
`;
