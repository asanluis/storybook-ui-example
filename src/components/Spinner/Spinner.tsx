import PropTypes from "prop-types";
import React from "react";
import {SpinnerProps} from "./Spinner.types";
import {CircleContainer, OneCircle} from "./Spinner.styles";

export const Spinner = (properties: SpinnerProps) => {

    const circleProps = Array.from(Array(12), (val, i) => ({
        rotate: `${properties.rotate * i}deg`,
        delay: `${-1.2 + i * properties.delay}s`
    }));
    return (
        <CircleContainer>
            {circleProps.map(circleProp => (
                <OneCircle key={circleProp.rotate} {...circleProp} />
            ))}
        </CircleContainer>
    );

}


Spinner.displayName = 'Spinner';

Spinner.propTypes = {
    rotate: PropTypes.number,
    delay: PropTypes.number,
};

export default Spinner;