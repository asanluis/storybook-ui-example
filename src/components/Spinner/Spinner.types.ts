export interface SpinnerProps {
    rotate: number
    delay: number
}