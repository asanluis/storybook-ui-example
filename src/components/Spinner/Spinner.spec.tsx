import React from 'react';
import {render} from "@testing-library/react";
import "@testing-library/jest-dom";
import Spinner from './index';
import {SpinnerProps} from "./Spinner.types";
import {ThemeProvider} from "styled-components";
import {lightTheme} from "../../themes/light.theme";


describe('[Spinner]', () => {
    it('Should render a Spinner', () => {
        const props: SpinnerProps = {
            rotate: 30, delay: 0.2
        }
        const {container} = render(<ThemeProvider theme={lightTheme}><Spinner {...props} /></ThemeProvider>);
        expect(container.firstChild).toMatchSnapshot();
    });
});

