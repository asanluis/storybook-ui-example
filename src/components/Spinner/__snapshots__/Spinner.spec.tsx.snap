// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Spinner] Should render a Spinner 1`] = `
<div
  class="sc-bcXHqe dHfpBF"
>
  <div
    class="sc-gswNZR inlFky"
    rotate="0deg"
  />
  <div
    class="sc-gswNZR gfLQxZ"
    rotate="30deg"
  />
  <div
    class="sc-gswNZR fBvYlF"
    rotate="60deg"
  />
  <div
    class="sc-gswNZR emWMsU"
    rotate="90deg"
  />
  <div
    class="sc-gswNZR dlLvPg"
    rotate="120deg"
  />
  <div
    class="sc-gswNZR ZoEIX"
    rotate="150deg"
  />
  <div
    class="sc-gswNZR cdsyeN"
    rotate="180deg"
  />
  <div
    class="sc-gswNZR izPPAE"
    rotate="210deg"
  />
  <div
    class="sc-gswNZR fyRNSE"
    rotate="240deg"
  />
  <div
    class="sc-gswNZR iQyZEK"
    rotate="270deg"
  />
  <div
    class="sc-gswNZR yFDYD"
    rotate="300deg"
  />
  <div
    class="sc-gswNZR gizFMf"
    rotate="330deg"
  />
</div>
`;
